### ISSUE TITLE
Should we use E.G()?

### ISSUE THREAD

**Issue by Chris-Hibbert** [Agoric/agoric-sdk#2502]
## What is the Problem Being Solved?

We don't have agreement on whether to use or recommend `E.G()`. Some people think it would be beneficial, others that it would be confusing for users. 
   Michael Fig suggested that `E.get()` might be more mnemonic than `E.G()`.
   What were the arguments for `E.G()`? Do they continue to hold since we've dropped `E.M()` and `E.C()`?

Use of E.G() was rhetorically connected to the discussion about dropping tildot, but is pretty much separate, since tildot doesn't support destructuring assignments.

This issue is a place to discuss pros and cons of allowing/encouraging use of `E.G()`.

## Description of the Design

`E.G()` makes it possible to do a destructuring assignment on an eventual value without waiting for the resolution of the value. Our infrastructure doesn't currently support pipelining, but once it does, this can be an additional opportunity for parallelism in our computational model.

## Security Considerations

misunderstanding asynchrony can be a source of errors.

I don't want to argue that adding or removing `await` statements is a security issue, but it is a performance issue, and can be a correctness issue.

## Test Plan

TBD once we have an agreement on how to proceed.

## The case against

This is not my position, so I'm tying to repeat what @dtribble  has said. The syntax for `E(foo).bar(something)` and `E.G(a)` are enough unlike the rest of javascript to be a learning barrier for the users we want to reach. In the absence of native syntax (tildot), we're stuck with `E()`, but we don't need `E.G()`, so we should do without it.

## The case in favor

When the result of an eventual message send must be destructured, but the results will be passed as parameters or used as receivers of further messages, adding await in order to do the destructuring is a change in semantics, and an unnecessary imposition on the control flow. I use `deposit()` in the following as an example of a function that requires its parameter to be resolved.

```js
const { x: xP, yP } = E.G(b.foo(c));
const r = yP.then(y => {
  return E(xP).deposit(y);
}
```

is more natural. Having to write without `E.G()` leads to 
```js
const { x: xP, y: yP } = await b.foo(c);
const y = await yP;
const r = E(xP).deposit(y);
```

use of top-level await produces the following, which some prefer.
```js
const { x: xP, y: yP } = E.G(b.foo(c));
const y = await yP;
const r = E(xP).deposit(y)
```

Would those who favor of `E.G()` like to add more examples of things that are easier with, or not possible without it? What other arguments on either side have I missed?

**Comment by katelynsills** [comment]
I prefer the functionality of `E.G`, but I don't like the name. This may be impossible, but would it be possible to include the functionality in `E()`? It seems like we should be able to know whether something is a method or a property without the code specifying `E` or `E.G`

**Comment by Chris-Hibbert** [comment]
> I prefer the functionality of E.G, but I don't like the name. This may be impossible, but would it be possible to include the functionality in E()? 

Someone else will have to respond to your question. I don't know the answer. 

Are you happier with `E.get()` than `E.G()`?

**Comment by katelynsills** [comment]
> Are you happier with E.get() than E.G()?

Yeah, that is an improvement

**Comment by kriskowal** [comment]
I also prefer `E.get`. Q had `Q.get` and `promise.get` for analogous purpose. I expect this to eventually be `Promise.get` when it approaches the standard, if it’s palatable at all.

**Comment by erights** [comment]
> I prefer the functionality of `E.G`, but I don't like the name. This may be impossible, but would it be possible to include the functionality in `E()`? It seems like we should be able to know whether something is a method or a property without the code specifying `E` or `E.G`

It is impossible to do it reliably. Consider `E(p).foo` vs `E(p).foo()` when we don't know what `p` is or will be. Both cases have `E(p).foo` as a subexpression. The proxy produced by `E(p)` has to return something for the expression `E(p).foo` without knowing whether this is an eventual method call or an eventual property access. In order to be able to follow it with `()` and have it mean what we want, `E(p).foo` much be callable, i.e., it must be a function. But if `E(p).foo` is a property access, then it must return a promise for the value of the property. 

It is impossible for an object to be both a function and a promise. It is possible for an object to be both a function and a thenable, which will act promise-like to some degree. But that way madness lies.

**Comment by erights** [comment]
I'm happy with the name `E.get` as in `E.get(p).foo`. I agree that this is more teachable and memorable than `E.G(p).foo`. It is also easier to partially guess its meaning if you're reading code and only know the meaning of `E(p).foo()`

**Comment by kriskowal** [comment]
> I also prefer `E.get`. Q had `Q.get` and `promise.get` for analogous purpose. I expect this to eventually be `Promise.get` when it approaches the standard, if it’s palatable at all.

I should clarify. Q had `Q.get(Promise, string) => Promise` and `Promise.prototype.get(string) => Promise`. My intuition is that the latter and `Promise.get(Promise, string) => Promise` are the path of least resistance for the standard.

**Comment by erights** [comment]
Then you'd need to write `Promise.resolve(p).get('foo')`. You can't use that `get` to enable you to write `.foo` so it looks like a property name. Surprise bonus is that @michaelfig was able to teach typescript to type these property accesses because it already had a theory of separately typed properties.

OTOH, I would not want to propose to the committee that something proxy based, like `E(p)` or `E.get(p)` be added to `Promise`. I think it is cleaner as a distinct helper. But I'll admit that may just be because I'm used to it.

**Comment by erights** [comment]
> ```js
> const { x: xP, yP } = E.G(b.foo(c));
> const r = yP.then(y => {
>   return E(xP).deposit(y);
> }
>
> is more natural. 

It is more natural. Unfortunately `.then` is not properly defensive ( #88 ). It just so happens this use of `.then` is not vulnerable to #88 because `E.G` will reliably bind `xP` to a fresh promise. However, that's more delicate reasoning than we'd like to teach anyone. That's why  we invented `E.when`, to have a safe defensive alternative to `.then`. Using it in your example produces

```js
const { x: xP, yP } = E.G(b.foo(c));
const r = E.when(yP, y => {
  return E(xP).deposit(y);
});
```
which is less natural because it is not what people are already familiar with. If there is any way we can make `.then` defensive enough without forking the language, we should. But I doubt there is. In any case, `.then` vs `E.when` is mostly besides the point of this issue.

> Having to write without `E.G()` leads to
>
> ```js
> const { x: xP, y: yP } = await b.foo(c);
> const y = await yP;
> const r = E(xP).deposit(y);
> ```

There's no need for a double await. That can be just

```js
const { x, y } = await b.foo(c);
const r = E(x).deposit(y);
```

> use of top-level await produces the following, which some prefer.

Your previous example had two top-level awaits and no other kind of await.

> 
> ```js
> const { x: xP, y: yP } = E.G(b.foo(c));
> const y = await yP;
> const r = E(xP).deposit(y)
> ```

That is a correct alternative to your first `then`ish code.

> Would those who favor of `E.G()` like to add more examples of things that are easier with, or not possible without it? What other arguments on either side have I missed?

**Comment by erights** [comment]
> Would those who favor of E.G() like to add more examples of things that are easier with, or not possible without it? What other arguments on either side have I missed?

Aside from brevity and clarity, to get all the benefits of
```js
const { x: xP, yP } = E.get(b.foo(c));
```
without `E.get` you'd have to write something like
```js
const pairP = b.foo(c);
const xP = E.when(pairP, pair => pair.x);
const yP = E.when(pairP, pair => pair.y);
```
All the alternatives above without `E.get` are semantically worse. They introduce an `await`. The code above retains all the semantic benefits but does so at a syntactic cost in verbosity and obscurity that seems terrible to me.

**Comment by kriskowal** [comment]
To be clear on my earlier points, given that E is analogous to Promise and
its nature is to provide proxies, I think E.get returning an accessor proxy
is coherent with an implied lower level Promise.get that returns a Promise.

That is to say, I like E.get as proposed both in name and nature.

**Comment by Chris-Hibbert** [comment]
There are methods that return a promise for a payment. `deposit()` doesn't accept unresolved values, so there are cases where the second await is necessary.

```
const { x: xP, y: yP } = await b.foo(c);
const y = await yP;
const r = E(xP).deposit(y);
```
is not always the same as 
```
const { x, y } = await b.foo(c);
const r = E(x).deposit(y);
```

`seat.getPayouts()` returns promises for payouts.

The `E.get()` form is much better here.

**Comment by erights** [comment]
> is not always the same as

You are correct. I overlooked the case of fulfilling to record containing promises. Thanks.

**Comment by Chris-Hibbert** [comment]
We had a discussion during the eng sync today. I think the conclusion was that with the decision to drop @agoric/babel-parser (in order to be able to upgrade to get more complete support of bigInts), we should move forward with `E.get()`. 

There's still some sensitivity around use of `E.get()` in destructuring assignments. I think the concern is that `E.get(expr)` doesn't read like an expression that should let you reach inside a promise and extract individual fields. When writing these complex expressions, compare versions with and without the destructuring and look for less complex ways to express the intent.

**Comment by michaelfig** [comment]
I don't think anybody expressed what seems to me the most direct way of doing @Chris-Hibbert's example (`E.get` without destructuring):

```js
const xyP = b.foo(c);
const xP = E.get(xyP).x;
const y = await E.get(xyP).y;
const r = E(xP).deposit(y);
```

This is one line longer than the destructured `E.get`, with the same semantics.

As @dtribble said, though, having to write such code is an API smell.

**Comment by katelynsills** [comment]
> As @dtribble said, though, having to write such code is an API smell.

I want to try to explain the constraints for some of our API decisions, since I don't think they are widely understood. `E(purse).deposit` requires a payment, not a promise for a payment, as per ordering requirements as determined by @erights. This means that although we accept promises in nearly everywhere else in our APIs, we cannot accept promises in the case of depositing a payout.

**Comment by michaelfig** [comment]
> > As @dtribble said, though, having to write such code is an API smell.
>
> This means that although we accept promises in nearly everywhere else in our APIs, we cannot accept promises in the case of depositing a payout.

I understand that.  What I mean is requiring `E.get` to manipulate a promise for a record that needs to be broken apart is a code smell.  A better API is something like the `UserSeat` presence, which exposes other pieces via getters that don't require `E.get` since plain `E` can be used.

**Comment by dtribble** [comment]
> > As @dtribble said, though, having to write such code is an API smell.
> 
> I want to try to explain the constraints for some of our API decisions, since I don't think they are widely understood. 

Just to be clear, a code smell is just a flag to reconsider. The answer might still be "yep this one is an exception and here's why".  So having to dance around for `deposit` might be entirely reasonable. In this case, as @michaelfig notes, it's the combination with the struct return that creates the pipelining issue. That too might be reasonable, but it's a signal that it's something to occasionally revisit as we get more clarity on usage patterns.

**Comment by katelynsills** [comment]
Sure, but it's equally important to know the constraints when mentioning potential code smell, as with more knowledge, it may be a false alarm. Without knowing the constraints, it's easy to think that we should just "fix" certain aspects, and wonder why we haven't. So, here's a quick list of the constraints:

1) `deposit` must take a payment, not a promise for a payment.
2) Zoe cannot wait for withdrawals as that makes it vulnerable to malicious/buggy issuers and violates payout liveness. This means it must expose a promise for a payment to users. 
3) The user will get payouts it did not expect, so the user must be able to receive all the payouts as a collection.

Note that the API currently has two methods for getting payouts, one for getting a promise for a payment separately if the keyword is known, and another for getting the payouts as a record:

```js
const promiseForPayment = E(seat).getPayout('Asset');
const { Asset: promiseForPayment, ... } = E(seat).getPayouts();
```
4) We must know which purse to deposit each payment into. The brand of the purse must match the brand of the payment, and moreover, the user must choose which purse if there are several purses with the same brand.
5) We cannot pass purses to Zoe, as purses should be closely held. 
6) Keywords are the way that users track what payout is which, as payouts may be for the same brand.

**Comment by dtribble** [comment]
@katelynsills Completely agreed. This is a great list. Those are the key architectural requirements that drive the API, and I agree with all of them. (At different points in time I remember an overlapping subset of these :). @rowgraus @erights These should (and may already be) be captured into architecture doc.

**Comment by dtribble** [comment]
> 3\. The user will get payouts it did not expect, so the user must be able to receive all the payouts as a collection.

I note that destructuring assignment actually impairs this requirement, since you will only get the subset of properties the user was expecting. It would be helpful to have a pattern or helper that provided an easy way to check for any payments that had not been processed. With destructuring that would be "have a `...rest` property and consume it or check it for empty". Right now taht just drops on the floor and we don't notice. This needs some design consideration because those assets are not recoverable. I'm especially concerned with cases where the user was given "change".

**Comment by dtribble** [comment]
Somehow my long response to this got lost.  I completely agree with Kate's comment. Sometimes a smell is just a smell, and just need to capture why it's appropriate. Usually those places are _interesting_. In this particular case, that list is GREAT and should be captured in an architecture doc.

**Comment by Chris-Hibbert** [comment]
The conclusion is that `E.get()` (as it has been renamed) is good and useful. We've merged #2616, which converts existing calls to `E.G()` to the new spelling.
